hip/components/memory.hip.hpp

```diff
@@ -20,71 +21,203 @@ namespace kernels {
+    static_assert(sizeof(atomic_type) == sizeof(ValueType), "invalid map");
+    static_assert(alignof(atomic_type) == sizeof(ValueType), "invalid map");
+    atomic_type cast_value{};
+    memcpy(&cast_value, &value, sizeof(ValueType));
+    HIP_ATOMIC_STORE(reinterpret_cast<atomic_type*>(ptr), cast_value, memorder,
+                     scope);
+}
+
+
+template <typename ValueType,
+          gcc_atomic_intrinsic_type_map<ValueType>* = nullptr>
 __device__ __forceinline__ ValueType load_relaxed(const ValueType* ptr)
 {
-    return load(ptr, 0);
+    return load_generic<__ATOMIC_RELAXED, HIP_SCOPE_GPU>(ptr);
 }
 
 
-template <typename ValueType>
+template <typename ValueType,
+          gcc_atomic_intrinsic_type_map<ValueType>* = nullptr>
+__device__ __forceinline__ ValueType load_relaxed_shared(const ValueType* ptr)
+{
+    return load_generic<__ATOMIC_RELAXED, HIP_SCOPE_THREADBLOCK>(ptr);
+}
+
+
+template <typename ValueType,
+          gcc_atomic_intrinsic_type_map<ValueType>* = nullptr>
 __device__ __forceinline__ ValueType load_acquire(const ValueType* ptr)
 {
-    auto result = load(ptr, 0);
-    __threadfence();
-    return result;
+    return load_generic<__ATOMIC_ACQUIRE, HIP_SCOPE_GPU>(ptr);
 }
 
-template <typename ValueType>
+
+template <typename ValueType,
+          gcc_atomic_intrinsic_type_map<ValueType>* = nullptr>
+__device__ __forceinline__ ValueType load_acquire_shared(const ValueType* ptr)
+{
+    return load_generic<__ATOMIC_ACQUIRE, HIP_SCOPE_THREADBLOCK>(ptr);
+}
+
+
+template <typename ValueType,
+          gcc_atomic_intrinsic_type_map<ValueType>* = nullptr>
 __device__ __forceinline__ void store_relaxed(ValueType* ptr, ValueType value)
 {
-    store(ptr, 0, value);
+    store_generic<__ATOMIC_RELAXED, HIP_SCOPE_GPU>(ptr, value);
 }
 
 
-template <typename ValueType>
+template <typename ValueType,
+          gcc_atomic_intrinsic_type_map<ValueType>* = nullptr>
+__device__ __forceinline__ void store_relaxed_shared(ValueType* ptr,
+                                                     ValueType value)
+{
+    store_generic<__ATOMIC_RELAXED, HIP_SCOPE_THREADBLOCK>(ptr, value);
+}
+
+
+template <typename ValueType,
+          gcc_atomic_intrinsic_type_map<ValueType>* = nullptr>
 __device__ __forceinline__ void store_release(ValueType* ptr, ValueType value)
 {
-    __threadfence();
-    store(ptr, 0, value);
+    store_generic<__ATOMIC_RELEASE, HIP_SCOPE_GPU>(ptr, value);
+}
+
+
+template <typename ValueType,
+          gcc_atomic_intrinsic_type_map<ValueType>* = nullptr>
+__device__ __forceinline__ void store_release_shared(ValueType* ptr,
+                                                     ValueType value)
+{
+    store_generic<__ATOMIC_RELEASE, HIP_SCOPE_THREADBLOCK>(ptr, value);
 }
 
 
 template <typename ValueType>
-__device__ __forceinline__ ValueType load_relaxed_shared(const ValueType* ptr)
+__device__ __forceinline__ thrust::complex<ValueType> load_relaxed(
```

Comment:
maybe only for thrust::complex<double>?
thrust::complex<float> can be done with int64

Reply:
I don't follow, we only use thrust::complex with float value types